icarusalg/Utilities/TrackTimeInterval.cxx

```diff
@@ -0,0 +1,318 @@
+   *
+   *     t(P) = t(ref) + T (R(P) - R(ref))
+   *     t0 = t(P) - dX / v = t(ref) + T (R(P) - R(ref)) - dX / v
+   *
+   * We don't know what dX is and the point of this function is to return the
+   * allowed interval of t0, which is covered by all the values of dX between
+   * `0` (on the anode plane) and `D` (distance of the cathode from the anode
+   * plane):
+   * 
+   *     t0(max) = t(ref) + T (R(P) - R(ref))
+   *     t0 in [ t0(max) - D / v ; t0(max) ]
+   *
+   * t(ref) is the time of the trigger, so if we express t0 relative to the
+   * trigger time we can omit it. The electronics time, on the other end,
+   * does not refer to the trigger time.
+   * `DetectorTimings` understands the time scale of R(P) (TPC electronics tick)
+   * and can convert it to other scales, e.g. the "standard" electronics scale,
+   * and give directly t0(max) in those other scales.
+   */
+  
+  detinfo::timescales::electronics_time const time
+    = fDetTimings.toElectronicsTime(TPCtick);
+  
+  TimeLimits_t const& planeTimeLimits = fGeomCache.limits[planeID];
+  microseconds const driftTime{
+    planeTimeLimits.driftDistance.convertInto<centimeters>().value()
+    / fDriftVelocity
+    };
+  return { time - driftTime, time };
+  
+} // lar::util::TrackTimeInterval::timeRange(tick)
+
+
+// -----------------------------------------------------------------------------
+lar::util::TrackTimeInterval::TrackTimeInterval(
+  GeometryCache_t geomCache,
+  detinfo::DetectorPropertiesData detProp,
+  detinfo::DetectorTimings detTimings
+)
+  : fDetProp{ std::move(detProp) }
+  , fDetTimings{ std::move(detTimings) }
+  , fDriftVelocity{ fDetProp.DriftVelocity() }
+  , fGeomCache{ std::move(geomCache) }
+  {}
+
+
+// -----------------------------------------------------------------------------
+auto lar::util::TrackTimeInterval::buildGeomCache(geo::GeometryCore const& geom)
+  -> GeometryCache_t
+{
+  return {
+      extractTimeLimits(geom),                   // limits
+      { geom.Ncryostats(), geom.MaxTPCsets() },  // TPCsetDims
+      extractTPCtoSetMap(geom)                   // TPCtoSet
+    };
+} // lar::util::TrackTimeInterval::buildGeomCache()
+
+
+// -----------------------------------------------------------------------------
+auto lar::util::TrackTimeInterval::extractTimeLimits
+  (geo::GeometryCore const& geom) -> GeometryCache_t::LimitsCache_t
+{
+  // this infrastructure and coding style is a remnant of a previous version
+  // which needed more information per plane and that information was different
+  // by plane; the following is still correct, has fewer assumptions and it's
+  // easier to expand, but a bit more wordy.
+  GeometryCache_t::LimitsCache_t limits = geom.makePlaneData<TimeLimits_t>();
+  
+  for (geo::TPCGeo const& TPC: geom.Iterate<geo::TPCGeo>()) {
+    geo::PlaneGeo const& firstPlane = TPC.FirstPlane();
+    
+    // in ICARUS the hit time is corrected as if it were on the first plane
```

Comment:
Actually this is something I'm unsure of: in the data I think this is the case, but there are some uncertainties with offsets and planes in the 2d sim, 1d signal processing step with ICARUS at the moment. (Thinking about the discussion last week with vertex X residual plots in MC)

Maybe doesn't matter enough to hold up this PR, but worth thinking about.

Reply:
Definitely good to know (not that I do, yet).
This utility is on the simple side, and I do expect people to use it with ample margin given the general uncertainty on the hit times vs. _x_ (for example because of field non-uniformity and cathode deformations).
This does not excuse me from providing the best possible tool.